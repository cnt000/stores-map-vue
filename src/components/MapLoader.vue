<template>
  <div>
    <template v-if="!!mapLoaded">
      <slot :google="google" :map="map" />
    </template>
    <div class="map-container">
      <div id="map" class="map"></div>
    </div>
  </div>
</template>

<script>
import GoogleMapsApiLoader from "google-maps-api-loader";
import { mapState, mapActions } from "vuex";

export default {
  props: {
    mapConfig: Object,
    apiKey: String
  },
  data() {
    return {
      google: null,
      geocoder: null
    };
  },
  mounted() {
    GoogleMapsApiLoader({
      apiKey: this.apiKey
    }).then(google => {
      this.google = google;
      this.initializeMap();
    });
  },
  computed: {
    ...mapState({
      store: state => state.stores.selectedId,
      mapLoaded: state => state.stores.mapLoaded,
      filters: state => state.stores.filters,
      keyword: state => state.stores.keyword
    })
  },
  watch: {
    store: function() {
      this.mapLoaded && this.panToSelectedStore();
    },
    mapLoaded: function() {
      this.store && this.panToSelectedStore();
    },
    filters: function() {
      this.filterStores();
    },
    keyword: function() {
      this.filterStores();
    }
  },
  methods: {
    ...mapActions("stores", [
      "filterStores",
      "getActive",
      "mapIsLoaded",
      "getSelectedStore"
    ]),
    initializeMap() {
      const mapContainer = this.$el.querySelector("#map");
      const { Map, Geocoder } = this.google.maps;
      this.map = new Map(mapContainer, this.mapConfig);
      this.google.maps.event.addListener(this.map, "idle", this.boundsChanged);
      this.geocoder = new Geocoder();
      this.mapIsLoaded(this.map);
    },
    boundsChanged() {
      this.getActive(this.map);
      this.filterStores();
    },
    panToSelectedStore() {
      const storeSelected = this.$store.getters["stores/getStore"];
      let zoom = storeSelected.id ? 15 : 10;
      this.panToAndZoom(storeSelected.lat, storeSelected.lng, zoom);
    },
    panToAndZoom(lat, lng, zoom) {
      this.map.panTo({
        lat: parseFloat(lat),
        lng: parseFloat(lng)
      });
      const actualZoom = this.map.getZoom();
      if (actualZoom < zoom) this.map.setZoom(zoom);
    }
  }
};
</script>

<style lang="scss">
.app-container {
  display: grid;
  grid-template-areas: "nav content";
  grid-template-columns: 300px 1fr;
  grid-template-rows: auto 1fr;
  grid-gap: 2px;
}
.map-container {
  grid-area: content;
  height: 100vh;
}
.map {
  height: 100%;
}
</style>
