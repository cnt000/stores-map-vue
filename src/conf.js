export const mapConfig = {
  zoom: 4,
  center: {
    lat: 51,
    lng: 0
  }
};

export const apiKey = "AIzaSyBNzPxDEDzlMCA9cedItIPCwtbdk037BGg";
