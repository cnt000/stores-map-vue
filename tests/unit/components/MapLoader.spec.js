import { mount, shallowMount, createLocalVue } from "@vue/test-utils";
import Vuex from "vuex";
import MapLoader from "@/components/MapLoader.vue";
import storesExamples from "../stores-examples.json";
import countriesExamples from "../countries-examples.json";
import { mapConfig, apiKey } from "@/conf";

const localVue = createLocalVue();
localVue.use(Vuex);

describe("MapLoader.vue", () => {
  let actions;
  let store;
  let testState;
  let GoogleMapsApiLoader;
  let initializeMap;

  beforeEach(() => {
    actions = {};

    testState = {
      all: storesExamples,
      pending: false,
      error: false,
      selectedStoreId: 0,
      countries: countriesExamples,
      selectedCountryId: 0
    };

    store = new Vuex.Store({
      modules: {
        stores: {
          state: testState,
          actions
        }
      }
    });

    GoogleMapsApiLoader = jest.fn();
    initializeMap = jest.fn();
  });

  it("matches snapshot", () => {
    const wrapper = mount(MapLoader, {
      store,
      localVue
    });
    expect(wrapper.html()).toMatchSnapshot();
  });

  it("create maps div", () => {
    const wrapper = shallowMount(MapLoader, {
      store,
      localVue
    });
    expect(wrapper.find("div#map").html()).toMatch(
      '<div id="map" class="map"></div>'
    );
  });

  it("has props apiKey and mapConfig object", () => {
    const wrapper = mount(MapLoader, {
      propsData: {
        apiKey: "GAPIkey",
        mapConfig: { a: 1, b: 3 }
      },
      store,
      localVue
    });
    console.log(wrapper.html());
    expect(wrapper.props().apiKey).toBe("GAPIkey");
    expect(wrapper.props().mapConfig).toEqual({ a: 1, b: 3 });
  });
});

/* 
GoogleMapsApiLoader, carica le mappe? e carica le initializeMap
initializeMap, fa mapIsLoaded?
mapIsLoaded, dispatcha?
panToSelectedStore: con lat e lon, fa panToAndZoom?
con storeAddress, fa geocoder.geocode?
selectCountry, con selectCountryId, fa panToAndZoom
se selectCountry != 0, fa geocode di countrySelected string?
panToAndZoom, fa panTo e zoom?
*/
