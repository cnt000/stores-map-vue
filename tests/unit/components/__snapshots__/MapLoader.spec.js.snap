// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MapLoader.vue matches snapshot 1`] = `
<div>
  <!---->
  <div class="map-container">
    <div id="map" class="map"></div>
  </div>
</div>
`;
